pkg/liquidity-mining/contracts/admin/GaugeAdder.sol

```diff
@@ -43,143 +54,147 @@ contract GaugeAdder is IGaugeAdder, SingletonAuthentication, ReentrancyGuard {
+            IERC20 pool = IStakingLiquidityGauge(gauge).lp_token();
+            require(pool != _balWethBpt, "Cannot add gauge for 80/20 BAL-WETH BPT");
+        }
+
+        _addGauge(gauge, gaugeTypeBytes);
+    }
+
+    /// @inheritdoc IGaugeAdder
+    function setGaugeFactory(ILiquidityGaugeFactory factory, string memory gaugeType) external override authenticate {
+        bytes32 gaugeTypeBytes = _validateAndCastGaugeType(gaugeType);
 
         // Sanity check that calling `isGaugeFromFactory` won't revert
         require(!factory.isGaugeFromFactory(address(0)), "Invalid factory implementation");
 
-        EnumerableSet.AddressSet storage gaugeFactories = _gaugeFactoriesByType[gaugeType];
-        require(gaugeFactories.add(address(factory)), "Factory already added");
+        _gaugeTypeFactory[gaugeTypeBytes] = factory;
 
-        emit GaugeFactoryAdded(gaugeType, factory);
+        emit GaugeFactorySet(gaugeType, gaugeType, factory);
     }
 
     // Internal functions
 
+    function _isGaugeFromValidFactory(address gauge, bytes32 gaugeType) internal view returns (bool) {
+        ILiquidityGaugeFactory gaugeFactory = _gaugeTypeFactory[gaugeType];
+        return gaugeFactory == ILiquidityGaugeFactory(0) ? false : gaugeFactory.isGaugeFromFactory(gauge);
+    }
+
     /**
      * @dev Adds `gauge` to the GaugeController with type `gaugeType` and an initial weight of zero
      */
-    function _addGauge(address gauge, GaugeType gaugeType) private {
-        require(isGaugeFromValidFactory(gauge, gaugeType), "Invalid gauge");
+    function _addGauge(address gauge, bytes32 gaugeType) private {
+        require(_isGaugeFromValidFactory(gauge, gaugeType), "Invalid gauge");
+        int128 gaugeTypeNumber = _gaugeTypeNumber[gaugeType];
 
         // `_gaugeController` enforces that duplicate gauges may not be added so we do not need to check here.
         _authorizerAdaptorEntrypoint.performAction(
             address(_gaugeController),
-            abi.encodeWithSelector(IGaugeController.add_gauge.selector, gauge, gaugeType)
+            abi.encodeWithSelector(IGaugeController.add_gauge.selector, gauge, gaugeTypeNumber)
         );
     }
+
+    function _stringToBytes32(string memory str) internal pure returns (bytes32 bytesString) {
+        uint256 length = bytes(str).length;
+        require(length > 0 && bytes(str).length <= 32, "Input string should be between 1 and 32 characters long");
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            bytesString := mload(add(str, 32))
+        }
+    }
+
+    function _bytes32ToString(bytes32 _bytes32) internal pure returns (string memory) {
+        uint256 length = 0;
+        while (length < 32 && _bytes32[length] != 0) {
+            ++length;
+        }
+
+        bytes memory byteArray = new bytes(length);
+        for (uint256 i = 0; i < length; ++i) {
+            byteArray[i] = _bytes32[i];
+        }
+
+        return string(byteArray);
```

Comment:
This part can be done in assembly as well. I don't know what's worse in this case :/